### PR TITLE
Add compensation integration

### PR DESCRIPTION
**Pull request by Petro31**
Adds the compensation integration

## Proposed change

Adds the compensation integration documentation

## Type of change

- [ ] Spelling, grammar or other readability improvements (`current` branch).
- [ ] Adjusted missing or incorrect information in the current documentation (`current` branch).
- [x] Added documentation for a new integration I'm adding to Home Assistant (`next` branch).
  - [ ] I've opened up a PR to add logo's and icons in [Brands repository](https://github.com/home-assistant/brands).
- [ ] Added documentation for a new feature I'm adding to Home Assistant (`next` branch).
- [ ] Removed stale or deprecated documentation.

## Additional information

- Link to parent pull request in the codebase: https://github.com/home-assistant/core/pull/41675
- Link to parent pull request in the Brands repository: 
- This PR fixes or closes issue: 

## Checklist

- [x] This PR uses the correct branch, based on one of the following:
  - I made a change to the existing documentation and used the `current` branch.
  - I made a change that is related to an upcoming version of Home Assistant and used the `next` branch.
- [x] The documentation follows the Home Assistant documentation [standards][].

[standards]: https://developers.home-assistant.io/docs/documenting/standards

### CODE REVIEW

File: source/_integrations/compensation.markdown [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      - 1.0 -> 0.0
+```
+
+{% configuration %}
+entity_id:
+  description: The entity to monitor.
+  required: true
+  type: string
+data_points:
+  description: The collection of data point conversions with the format `uncompensated_value -> compensated_value`.  e.g. `1.0 -> 2.1`. The number of required data points is equal to the polynomial `degree` + 1. For example, a linear compensation (with `degree: 1`) requires at least 2 data points.
```
Comment (klaasnicolaas) [inline source/_integrations/compensation.markdown]:
```suggestion
  description: "The collection of data point conversions with the format `uncompensated_value -> compensated_value`.  e.g. `1.0 -> 2.1`. The number of required data points is equal to the polynomial `degree` + 1. For example, a linear compensation (with `degree: 1`) requires at least 2 data points."
```

**Review by klaasnicolaas — DISMISSED** [review]

File: source/_integrations/compensation.markdown [hunk 2]
```diff
@@ -0,0 +1,66 @@
+---
+title: Compensation
+description: Instructions on how to integrate compensation sensors into Home Assistant.
+ha_category:
+  - Utility
+  - Sensor
+ha_iot_class: Local Push
```
Comment (frenck) [inline source/_integrations/compensation.markdown]: This integration calculates its states based on existing data.

```suggestion
ha_iot_class: Calculated
```

File: source/_integrations/compensation.markdown [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+ha_release: '0.117'
+ha_quality_scale: internal
+ha_codeowners:
+  - '@petro31'
+ha_domain: compensation
+---
+
+The `compensation` sensor platform consumes the state from other sensors. It exports the `compensated` value as state and the following values as attributes: `entity_id` and `coefficients`.
```
Comment (frenck) [inline source/_integrations/compensation.markdown]:
```suggestion
The Compensation integration consumes the state from other sensors. It exports the compensated value as state and the following values as attributes: `entity_id` and `coefficients`.
```

File: source/_integrations/compensation.markdown [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+To enable the compensation sensor, add the following lines to your `configuration.yaml`:
+
+```yaml
+# Example configuration.yaml entry
+sensor:
+  - platform: compensation
+    name: Yamaha Zone 1 Volume
+    entity_id: media_player.yamaha_receiver
+    attribute: volume_level
+    unit_of_measurement: dB
```
Comment (frenck) [inline source/_integrations/compensation.markdown]: We try to keep all configuration samples minimal. Thus, no optional requirement in the default sample. This helps a user to get started quickly by copy-&-paste the sample without worrying about optional parameters which they most likely not need. If required, insert a full configuration sample later that covers special setups or alike.

File: source/_integrations/compensation.markdown [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+The `compensation` sensor platform consumes the state from other sensors. It exports the `compensated` value as state and the following values as attributes: `entity_id` and `coefficients`.
+
+## Configuration
+
+To enable the compensation sensor, add the following lines to your `configuration.yaml`:
+
+```yaml
+# Example configuration.yaml entry
+sensor:
+  - platform: compensation
```
Comment (frenck) [inline source/_integrations/compensation.markdown]: We no longer accept YAML configuration for integrations on a platform level.

For more information about this decision see:

- ADR-0007: <https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md#decision>

File: source/_integrations/compensation.markdown [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  description: Name of the sensor to use in the frontend.
+  required: false
+  default: Compensation
+  type: string
+attribute:
+  description: Attribute to monitor.
+  required: false
+  type: string
+degree:
+  description: The degree of a polynomial. e.g. Linear compensation (y = x + 3) has 1 degree, Quadratic compensation (y = x<sup>2</sup> + x + 3) has 2 degrees, etc.
```
Comment (frenck) [inline source/_integrations/compensation.markdown]:
```suggestion
  description: "The degree of a polynomial. e.g., Linear compensation (`y = x + 3`) has 1 degree, Quadratic compensation (`y = x<sup>2</sup> + x + 3`) has 2 degrees, etc."
```

File: source/_integrations/compensation.markdown [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      - 1.0 -> 0.0
+```
+
+{% configuration %}
+entity_id:
+  description: The entity to monitor.
+  required: true
+  type: string
+data_points:
+  description: "The collection of data point conversions with the format `uncompensated_value -> compensated_value`.  e.g. `1.0 -> 2.1`. The number of required data points is equal to the polynomial `degree` + 1. For example, a linear compensation (with `degree: 1`) requires at least 2 data points."
```
Comment (frenck) [inline source/_integrations/compensation.markdown]:
```suggestion
  description: "The collection of data point conversions with the format `uncompensated_value -> compensated_value`.  e.g., `1.0 -> 2.1`. The number of required data points is equal to the polynomial `degree` + 1. For example, a linear compensation (with `degree: 1`) requires at least 2 data points."
```

**Review by frenck — CHANGES_REQUESTED** [review]
Thanks for the PR @Petro31.

I've reviewed it, and left some comments. Could you please take a look?

Thanks 👍

File: source/_integrations/compensation.markdown (hunk 2 above)
Comment (Petro31) [reply source/_integrations/compensation.markdown]: Yep, I assumed that was 'local push'.  Makes more sense as calculated.

File: source/_integrations/compensation.markdown (hunk 5 above)
Comment (Petro31) [reply source/_integrations/compensation.markdown]: Ok, I'll probably need to update the other PR to handle that.

File: source/_integrations/compensation.markdown (hunk 6 above)
Comment (Petro31) [reply source/_integrations/compensation.markdown]: will the `` kill the x<sup>2</sup>?  That's why I avoided it in the first place.

Comment (Petro31) [reply source/_integrations/compensation.markdown]: `test<sup>test</sup>` vs test<sup>test</sup>

Comment (Petro31) [reply source/_integrations/compensation.markdown]: Looks like it does

File: source/_integrations/compensation.markdown (hunk 4 above)
Comment (Petro31) [reply source/_integrations/compensation.markdown]: fixed

File: source/_integrations/compensation.markdown (hunk 5 above)
Comment (Petro31) [reply source/_integrations/compensation.markdown]: fixed

File: source/_integrations/compensation.markdown (hunk 6 above)
Comment (Petro31) [reply source/_integrations/compensation.markdown]: I left the equations as non-code to ensure the superscript stays.

**Comment by MatthewFlamm** [comment]
I think it is worth clearly stating that this integration does not interpolate.  It fits a single line (or higher order polynomial) to all the data points.  I think this will cause confusion as it does with the similar esphome usage.  For example:

```yaml
# Example configuration.yaml entry
compensation:
  media_player_db_volume:
    entity_id: media_player.yamaha_receiver
    data_points:
      - 0.2 -> -80.0
      - 0.5 -> -20.0
      - 1.0 -> -10.0
```

A value of 0.75 will not transform to -15.0.  In fact a value of 0.5 won't transform to -20 in this case either!

**Comment by Petro31** [comment]
> I think it is worth clearly stating that this integration does not interpolate. It fits a single line (or higher order polynomial) to all the data points. I think this will cause confusion as it does with the similar esphome usage. For example:
> 
> ```yaml
> # Example configuration.yaml entry
> compensation:
>   media_player_db_volume:
>     entity_id: media_player.yamaha_receiver
>     data_points:
>       - 0.2 -> -80.0
>       - 0.5 -> -20.0
>       - 1.0 -> -10.0
> ```
> 
> A value of 0.75 will not transform to -15.0. In fact a value of 0.5 won't transform to -20 in this case either!

@MatthewFlamm It's funny you should mention that.  I was planning on adding non-linear interpolation at some point down the road.  Does the ESPHome version do interpolation?

**Comment by MatthewFlamm** [comment]
I am only aware of the `calibrate_linear` (and `calibrate_polynomial`) filter in esphome that works like this integration.  I think an interpolate integration would also be useful, but this seems off topic.  My suggestion here is to add something like "A single polynomial, linear by default, is fit to all data points provided."

**Comment by Petro31** [comment]
> I am only aware of the `calibrate_linear` (and `calibrate_polynomial`) filter in esphome that works like this integration. I think an interpolate integration would also be useful, but this seems off topic. My suggestion here is to add something like "A single polynomial, linear by default, is fit to all data points provided."

I'm not the greatest tech writer, mind if I use that line?

**Comment by MatthewFlamm** [comment]
>  I'm not the greatest tech writer, mind if I use that line?

You can use it.

**Review by frenck — DISMISSED** [review]
✅ Approved. Can be merged as soon as the parent PR gets merged.

**Review by frenck — DISMISSED** [review]

File: source/_integrations/compensation.markdown [hunk 8]
```diff
@@ -0,0 +1,63 @@
+---
+title: Compensation
+description: Instructions on how to integrate compensation sensors into Home Assistant.
+ha_category:
+  - Utility
+  - Sensor
+ha_iot_class: Calculated
+ha_release: '0.117'
```
Comment (frenck) [inline source/_integrations/compensation.markdown]:
```suggestion
ha_release: "2021.4.0"
```

File: source/_integrations/compensation.markdown [hunk 9]
```diff
@@ -0,0 +1,62 @@
+---
+title: Compensation
+description: Instructions on how to integrate compensation sensors into Home Assistant.
+ha_category:
+  - Utility
+  - Sensor
+ha_iot_class: Calculated
+ha_release: "2021.4.0"
```
Comment (frenck) [inline source/_integrations/compensation.markdown]:
```suggestion
ha_release: "2021.5"
```

**Review by frenck — APPROVED** [review]
🎉 @Petro31 👍

**Pull request merged**